feature/amazon/src/main/java/com/revenuecat/purchases/amazon/handler/PurchaseHandler.kt

```diff
@@ -0,0 +1,120 @@
+import com.amazon.device.iap.model.PurchaseResponse
+import com.amazon.device.iap.model.Receipt
+import com.amazon.device.iap.model.RequestId
+import com.amazon.device.iap.model.UserData
+import com.revenuecat.purchases.ProductType
+import com.revenuecat.purchases.PurchasesError
+import com.revenuecat.purchases.PurchasesErrorCode
+import com.revenuecat.purchases.amazon.AmazonStrings
+import com.revenuecat.purchases.amazon.listener.PurchaseResponseListener
+import com.revenuecat.purchases.common.LogIntent
+import com.revenuecat.purchases.common.log
+import com.revenuecat.purchases.models.ProductDetails
+import com.revenuecat.purchases.strings.PurchaseStrings
+
+class PurchaseHandler : PurchaseResponseListener {
+
+    private val productTypes = mutableMapOf<String, ProductType>()
+    private val presentedOfferingsByProductIdentifier = mutableMapOf<String, String?>()
+    private val purchaseCallbacks =
+        mutableMapOf<RequestId, Pair<(Receipt, UserData) -> Unit, (PurchasesError) -> Unit>>()
+
+    override fun purchase(
+        appUserID: String,
+        productDetails: ProductDetails,
+        presentedOfferingIdentifier: String?,
+        onSuccess: (Receipt, UserData) -> Unit,
+        onError: (PurchasesError) -> Unit
+    ) {
+        log(LogIntent.PURCHASE, PurchaseStrings.PURCHASING_PRODUCT.format(productDetails.sku))
+
+        synchronized(this@PurchaseHandler) {
+            productTypes[productDetails.sku] = productDetails.type
+            presentedOfferingsByProductIdentifier[productDetails.sku] = presentedOfferingIdentifier
+        }
+
+        val requestId = PurchasingService.purchase(productDetails.sku)
+        purchaseCallbacks[requestId] = onSuccess to onError
+    }
+
+    override fun onPurchaseResponse(response: PurchaseResponse) {
+        log(LogIntent.DEBUG, AmazonStrings.PURCHASE_REQUEST_FINISHED.format(response.toJSON().toString(1)))
+
+        val requestId = response.requestId
+        val callbacks = purchaseCallbacks.remove(requestId)
+
+        callbacks?.let { (onSuccess, onError) ->
+            when (response.requestStatus) {
+                PurchaseResponse.RequestStatus.SUCCESSFUL ->
+                    onSuccessfulPurchase(response.receipt, response.userData, onSuccess)
+                PurchaseResponse.RequestStatus.FAILED -> onFailed(onError)
+                PurchaseResponse.RequestStatus.INVALID_SKU -> onInvalidSku(onError)
+                PurchaseResponse.RequestStatus.ALREADY_PURCHASED -> onAlreadyPurchased(onError)
+                PurchaseResponse.RequestStatus.NOT_SUPPORTED -> onNotSupported(onError)
+                null -> onUnknownError(onError)
+            }
+        }
+    }
+
+    @SuppressWarnings("ForbiddenComment")
+    private fun onSuccessfulPurchase(
+        receipt: Receipt,
+        userData: UserData,
+        onSuccess: (Receipt, UserData) -> Unit
+    ) {
+        if (receipt.isCanceled) {
+            // TODO: Do we do anything?
```

Comment:
This happens if the subscription is canceled or expired. I think this would only happen when restoring. What they do in the sample is to revoke the subscription in the local database. In our case, the server would be aware of this already and remove the entitlement.

I can see a case where in sandbox it would be valuable to inform the backend if this happens. Sandbox refreshes way slower than production. I will talk with @tonycosentini to see if it would be interesting to send the receipt if we detect the receipt has been canceled.